control-plane/agents/core/src/core/scheduling/mod.rs

```diff
@@ -0,0 +1,124 @@
+pub(crate) mod resources;
+pub(crate) mod volume;
+
+use crate::core::scheduling::{
+    resources::{PoolItem, ReplicaItem},
+    volume::GetSuitablePoolsContext,
+};
+use common_lib::types::v0::message_bus::PoolState;
+use std::{cmp::Ordering, collections::HashMap, future::Future};
+
+#[async_trait::async_trait(?Send)]
+pub(crate) trait ResourceFilter: Sized {
+    type Request;
+    type Item;
+
+    fn filter_iter(self, filter: fn(Self) -> Self) -> Self {
+        filter(self)
+    }
+    async fn filter_iter_async<F, Fut>(self, filter: F) -> Self
+    where
+        F: Fn(Self) -> Fut,
+        Fut: Future<Output = Self>,
+    {
+        filter(self).await
+    }
+    fn filter<F: FnMut(&Self::Request, &Self::Item) -> bool>(self, filter: F) -> Self;
+    async fn filter_async<Fn, Fut>(self, filter: Fn) -> Self
+    where
+        Fn: FnMut(&Self::Request, &Self::Item) -> Fut,
+        Fut: Future<Output = bool>;
+    fn sort<F: FnMut(&Self::Item, &Self::Item) -> std::cmp::Ordering>(self, sort: F) -> Self;
+    fn collect(self) -> Vec<Self::Item>;
+    fn group_by<K, V, F: Fn(&Self::Request, &Vec<Self::Item>) -> HashMap<K, V>>(
+        self,
+        _group: F,
+    ) -> HashMap<K, V> {
+        unimplemented!();
+    }
+}
+
+pub(crate) struct NodeFilters {}
+impl NodeFilters {
+    /// Should only attempt to use online nodes
+    pub(crate) fn online_nodes(_request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        item.node.is_online()
+    }
+    /// Should only attempt to use allowed nodes (by the topology)
+    pub(crate) fn allowed_nodes(request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        request.allowed_nodes().is_empty() || request.allowed_nodes().contains(&item.pool.node)
+    }
+    /// Should only attempt to use nodes not currently used by the volume
+    pub(crate) fn used_nodes(request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        let registry = &request.registry;
+        let used_nodes = registry.specs.get_volume_data_nodes(&request.uuid);
+        !used_nodes.contains(&item.pool.node)
+    }
+}
+
+pub(crate) struct PoolFilters {}
+impl PoolFilters {
+    /// Should only attempt to use pools with sufficient free space
+    pub(crate) fn enough_free_space(request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        item.pool.free_space() > request.size
+    }
+    /// Should only attempt to use healthy pools
+    pub(crate) fn healthy_pools(_: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        item.pool.state != PoolState::Faulted && item.pool.state != PoolState::Unknown
```

Comment:
What about degraded pools? If we are allowed to use these then we should change the function name from "healthy_pools" to something like "usable_pools".

Reply:
Done